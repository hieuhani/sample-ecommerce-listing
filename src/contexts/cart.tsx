import React, { useReducer, useContext, useMemo } from 'react'
import { Product } from '../repositories/products'

export interface CartProviderProps {
  children: React.ReactNode;
}

export interface Cart {
  [key: string]: {
    quantity: number;
    price: number;
  };
}

export interface State {
  cart: Cart;
}

export type Action =
  | { type: 'ADD'; payload: Product }
  | { type: 'DELETE'; payload: Product };

type Dispatch = (action: Action) => void

const initialState: State = {
  cart: {}
}

export type CartState = State & {
  totalItems: number;
  totalValue: number;
}

const CartStateContext = React.createContext<CartState>(initialState as CartState)
const CartDispatchContext = React.createContext<Dispatch | undefined>(undefined)

export function cartReducer(state: State, action: Action): State {
  switch (action.type) {
    case 'ADD': {
      const product = action.payload
      return {
        ...state,
        cart: Object.assign({}, state.cart, {
          [product.sku]: {
            price: product.price.sellPrice,
            quantity: state.cart[product.sku] ? state.cart[product.sku].quantity + 1 : 1,
          }
        })
      }
    }
    case 'DELETE': {
      const product = action.payload
      if (!state.cart[product.sku]) {
        return state
      }
      if (state.cart[product.sku].quantity > 1) {
        return {
          ...state,
          cart: Object.assign({}, state.cart, {
            [product.sku]: {
              price: product.price.sellPrice,
              quantity: state.cart[product.sku].quantity - 1,
            }
          })
        }
      }
      const { [product.sku]: value, ...cart } = state.cart
      return {
        ...state,
        cart,
      }
    }
  }
}

export function CartProvider({ children }: CartProviderProps): React.ReactNode {
  const [state, dispatch] = useReducer(cartReducer, initialState)
  const cartState: CartState = useMemo(() => {
    let totalItems = 0
    let totalValue = 0
    Object.keys(state.cart).forEach(productSku => {
      totalItems += state.cart[productSku].quantity
      totalValue += state.cart[productSku].quantity * state.cart[productSku].price
    })
    return {
      cart: state.cart,
      totalItems,
      totalValue,
    }
  }, [state])
  return (
    <CartStateContext.Provider value={cartState}>
      <CartDispatchContext.Provider value={dispatch}>
        {children}
      </CartDispatchContext.Provider>
    </CartStateContext.Provider>
  )
}

export function useCartState(): CartState {
  const context = useContext(CartStateContext)
  if (!context) {
    throw new Error('useCartState must be inside a CartProvider with a state value')
  }
  return context
}

export function useCartDispatch(): Dispatch {
  const context = useContext(CartDispatchContext)
  if (!context) {
    throw new Error('useCartDispatch must be inside a CartProvider with a state dispatch')
  }
  return context
}
