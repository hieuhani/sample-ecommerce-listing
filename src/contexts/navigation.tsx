import React, { useReducer, useContext, useMemo } from 'react'
import { ScreenPayload } from '../types'

export interface NavigationProviderProps {
  children: React.ReactNode;
}

export interface State {
  stack: Array<ScreenPayload>;
}

export type Action =
  | { type: 'PUSH'; payload: ScreenPayload }
  | { type: 'POP' };

type Dispatch = (action: Action) => void

const initialState: State = {
  stack: [{
    name: 'PRODUCT_LISTING',
  }]
}

export type ScreenPayloadWithOrder = ScreenPayload & {
  order: number;
}

export type NavigationState = State & {
  currentScreen: ScreenPayload | undefined;
  stackWithOrder: Array<ScreenPayloadWithOrder>;
}

const NavigationStateContext = React.createContext<NavigationState>(initialState as NavigationState)
const NavigationDispatchContext = React.createContext<Dispatch | undefined>(undefined)

export function navigationReducer(state: State, action: Action): State {
  switch (action.type) {
    case 'PUSH':
      return {
       ...state,
       stack: [...state.stack, action.payload],
      }
    case 'POP':
      return {
        ...state,
        stack: state.stack.length > 1 ? state.stack.filter((_, index) => index !== state.stack.length - 1) : initialState.stack
      }
    default:
      throw new Error()
  }
}

export function NavigationProvider({ children }: NavigationProviderProps): React.ReactNode {
  const [state, dispatch] = useReducer(navigationReducer, initialState)
  const navigationState: NavigationState = useMemo(() => {
    return {
      stack: state.stack,
      currentScreen: state.stack.length > 0 ? state.stack.slice(-1)[0] : undefined,
      stackWithOrder: state.stack.map((item, index) => Object.assign({}, item, {
        order: index,
      }))
    }
  }, [state])
  return (
    <NavigationStateContext.Provider value={navigationState}>
      <NavigationDispatchContext.Provider value={dispatch}>
        {children}
      </NavigationDispatchContext.Provider>
    </NavigationStateContext.Provider>
  )
}

export function useNavigationState(): NavigationState {
  const context = useContext(NavigationStateContext)
  if (!context) {
    throw new Error('useNavigationState must be inside a NavigationProvider with a state value')
  }
  return context
}

export function useNavigationDispatch(): Dispatch {
  const context = useContext(NavigationDispatchContext)
  if (!context) {
    throw new Error('useNavigationDispatch must be inside a NavigationProvider with a state dispatch')
  }
  return context
}
