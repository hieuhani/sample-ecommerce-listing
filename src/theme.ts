import { DefaultTheme } from 'styled-components'

export const theme: DefaultTheme = {
  colors: {
    reddishOrange: '#f5471e',
    coolGrey: '#8f9598',
    silver: '#c6cace',
    tomato: '#ea341f',
    darkGrey: '#262829',
    paleGrey: '#eef1f3',
    white: '#FFF',
    black30: 'rgba(0, 0, 0, .3)',
    pumpkinOrange: '#f58400',
    deepSkyBlue: '#1e75f6',
  }
}
