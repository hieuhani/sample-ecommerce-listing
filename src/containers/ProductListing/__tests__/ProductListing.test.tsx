import React from 'react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { ProductListing } from '..'
import { theme } from '../../../theme'
import { NavigationProvider } from '../../../contexts/navigation'

describe('SimilarProducts', () => {
  test('SimilarProducts renders correctly', () => {
    const tree = renderer.create(
      <ThemeProvider theme={theme}>
        <NavigationProvider>
          <ProductListing />
        </NavigationProvider>
      </ThemeProvider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
