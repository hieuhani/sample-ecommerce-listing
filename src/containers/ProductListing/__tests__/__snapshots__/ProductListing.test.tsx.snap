// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimilarProducts SimilarProducts renders correctly 1`] = `
Array [
  .c0 {
  height: 44pt;
  background: rgb(232,32,32);
  background: linear-gradient(90deg,rgba(232,32,32,1) 0%,#f5471e 100%);
  border-bottom: 1px solid #eef1f3;
  padding: 8pt 11pt;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  width: 24pt;
  height: 24pt;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-right: 11pt;
  border: none;
  background-color: transparent;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #FFF;
  border-radius: 8pt;
  height: 32pt;
  box-shadow: 0 2pt 4pt 0 rgba(0,0,0,.3);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 8pt;
}

.c4 {
  border: 0;
  width: 100%;
  height: 100%;
  border-radius: 8pt;
}

.c4:focus {
  outline: none;
}

.c3 {
  margin-right: 2pt;
}

<div
    className="c0"
  >
    <button
      className="c1"
      onClick={[Function]}
    >
      <img
        src="/public/images/ArrowBackWhite.png"
        srcSet="/public/images/ArrowBackWhite@2x.png 2x, /public/images/ArrowBackWhite@3x.png 3x"
      />
    </button>
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <img
          src="/public/images/Search.png"
          srcSet="/public/images/Search@2x.png 2x, /public/images/Search@3x.png 3x"
        />
      </div>
      <input
        className="c4"
        onChange={[Function]}
        placeholder="Nhập tên, mã sản phẩm"
      />
    </div>
  </div>,
  .c0 {
  overflow-y: auto;
  height: calc(100% - 44pt);
}

<div
    className="c0"
  />,
]
`;
